### PR TITLE
Add transformer that converts 4337 User Ops -> pseudo transactions

### PR DESCRIPTION
**Pull request by datadanne**
## Notes
* May need to add more accounts in the future
* We do support multiple user ops per transaction for the same account currently.
* handleAggregatedOps - if the bundler bundles multiple ops from different accounts into one root transaction, we don't support that

## Next steps
* Deploy this code in the API so that we can return pseudotransactions, with assetTransfers and net asset transfers for each.
* Generate test data for contextualizers.
* One new PR to consume pseudotransactions for a single tx and run context on it.
* Create new contextualizer for transactions that include user ops
* Outside of the repo, will need to call contextualizers on pseudotransactions. User op hash is returned in a field called meta
* Modify all to return new objects rather than passing by reference and modifying

## How it works
1) Split all User Operations from transactions sent to the ERC4337 entrypoint into separate 'UserOp transactions' (1 transaction to the entrypoint can include multiple user operations). This steps also removes log entries from the entry point contract and the eth transfer from the entry point to the bundler that submitted the transaction since those are not relevant to the user operation
2) Decode the call from the entry point to the smart account contract. Smart account contracts all contain functions that the entry point calls that then executes the user operation, so we need to decode that function call and get the arguments from that to get to what the user wants to do. There are lots of different account implementations where the entrypoint fn signature differ slightly but do the same thing:

   ```solidity
   function execute(address,uint256,bytes)                       // 0xb61d27f6
   function execute(address[],uint256[],bytes[])                 // 0xa516a5bf
   function execute(address,uint256,bytes,uint8)                 // 0x51945447
   function executeBatch((address,uint256,bytes)[])              // 0x34fcd5be
   function execTransactionFromEntrypoint(address,uint256,bytes) // 0x70641a22
   function executeBatch(address[],uint256[],bytes[])            // 0x47e1da2a
   ```

After these two steps the existing contextualizers can use the same logic as for non-ERC4337 transactions to give context to the transaction, I've included examples for a token approval, token transfer, eth transfer and an nft mint.

This WIP implementation only supports transactions where each user operation only contains "one action" (either just `execute` or `executeBatch` with just one action)

The diff is best viewed with whitespace hidden (https://github.com/Once-Upon/context/pull/291/files?diff=unified&w=1)

### CODE REVIEW

File: src/helpers/utils.ts [hunk 1]
```diff
@@ -196,12 +198,41 @@ export const convertDate = (epochTime: number): string => {
...
+
+export const isRawTransaction = (
+  v: PartialTransaction,
+): v is RawTransaction => {
+  return 'hash' in v;
+};
+
+const isTxnTransformer = (
+  v: TxnTransformer | BlockTransformer,
+): v is TxnTransformer => v.length == 2;
```
Comment (pcowgill) [inline src/helpers/utils.ts:216]: @datadanne Help me understand this length part?

File: src/transformers/_common/accountAbstraction/accounts/unknown.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import { Hex } from 'viem';
+import type { PseudoTransaction } from '../../../../types';
+import { decodeTransactionInput } from '../../../../helpers/utils';
+
+// TODO source unknown
```
Comment (pcowgill) [inline src/transformers/_common/accountAbstraction/accounts/unknown.ts:5]: @datadanne What's the back story on where this came from?

File: src/transformers/_common/accountAbstraction/index.spec.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    expect(txn?.pseudoTransactions).toBeDefined();
+    expect(txn?.pseudoTransactions?.length).toBe(1);
+
+    const pseudoTransaction = txn!.pseudoTransactions![0];
+
+    expect(pseudoTransaction).toMatchObject({
+      from: '0x2991c3845396c9f1d262b2ca0674111a59e2c90a',
+      to: '0x5d72015cc621025c265fabffc2fa55ac4821d79f',
+      receipt: expect.objectContaining({ status: 1 }),
+    });
```
Comment (pcowgill) [inline src/transformers/_common/accountAbstraction/index.spec.ts:31]: Having this tested is great

File: src/transformers/_common/accountAbstraction/index.spec.ts [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    // Check that the full transaction has logs from the entry point
+    expect(txn.receipt).toMatchObject({
+      logs: expect.arrayContaining([
+        expect.objectContaining({
+          address: ENTRY_POINT_V060,
+        }),
+      ]),
+    });
+
+    // Check that the pseudoTransaction does NOT have logs from the entry point
```
Comment (pcowgill) [inline src/transformers/_common/accountAbstraction/index.spec.ts:72]: This is great

File: src/transformers/_common/assetTransfers.ts [hunk 5]
```diff
@@ -173,87 +172,83 @@ function getTokenTransfers(tx: RawTransaction) {
...
+    assetTransfers.push(...leftOverTxfers);
+  }
 
-    return tx;
-  });
+  if (assetTransfers.length > 0) {
+    tx.assetTransfers = assetTransfers;
+  }
 
-  return block;
```
Comment (pcowgill) [inline src/transformers/_common/assetTransfers.ts:258]: @datadanne Note to self: Find where this functions is called to make sure it can still be used by passing in a block and getting a block back cc @runninyeti

Comment (pcowgill) [reply src/transformers/_common/assetTransfers.ts:258]: @datadanne Maybe we can also export an alternate version of this that can be applied to an entire block, and it uses this more targeted transform function under the hood?

So each of these files has one function that takes a block and returns a block, and it uses a function that takes a block and returns a transaction for each tx in that block?

File: src/transformers/_common/sigHash.ts [hunk 6]
```diff
@@ -1,19 +1,17 @@
...
-export function transform(block: RawBlock): RawBlock {
-  block.transactions = block.transactions.map((tx) => {
-    const sigHash = tx.input.slice(0, 10);
-    const internalSigHashes: InternalHashType[] = tx.traces.map((trace) => ({
-      from: trace.action.from,
-      to: trace.action.to,
-      sigHash: String(trace.action.input).slice(0, 10),
-    }));
+export const transform: TxnTransformer = (_block, tx) => {
+  if (!isRawTransaction(tx)) return tx;
```
Comment (pcowgill) [inline src/transformers/_common/sigHash.ts:5]: @datadanne Why was this needed for this one but not for the others?

**Review by pcowgill — APPROVED** [review]

File: src/transformers/_common/assetTransfers.ts [hunk 7]
```diff
@@ -173,87 +172,83 @@ function getTokenTransfers(tx: RawTransaction) {
   return txAssetTransfers;
 }
 
-export function transform(block: RawBlock): RawBlock {
-  block.transactions = block.transactions.map((tx) => {
-    // don't count transfers for failed txs
-    if (!tx.receipt.status || !tx.to) {
-      return tx;
-    }
+export const transform: TxnTransformer = (_block, tx) => {
```
Comment (pcowgill) [inline src/transformers/_common/assetTransfers.ts:175]: @datadanne To make sure I'm reading the ts properly, what is the type that's returned from this function (and all similar functions using this type)?

File: src/helpers/utils.ts (hunk 1 above)
Comment (datadanne) [reply src/helpers/utils.ts:216]: Ah, this isn't super obvious but `.length` on a value that is a function returns the number of arguments the function accepts. `TxnTransformer` accepts two arguments and `BlockTransformer` one

File: src/transformers/_common/assetTransfers.ts (hunk 7 above)
Comment (datadanne) [reply src/transformers/_common/assetTransfers.ts:175]: `TxnTransformer` is a generic function defined as
```ts
export type TxnTransformer = <T extends PartialTransaction>(
  block: RawBlock,
  transaction: T,
) => T;
```
so the value returned from the function depends on the input, it is implemented this way because we want to be able to transform both `RawTransaction`s and `PseudoTransaction`s. (In `makeTransform` it is called on `block.transactions` where the items have type `RawTransaction` and transaction.pseudoTransactions where the items have type `PseudoTransaction[]`)

File: src/transformers/_common/assetTransfers.ts (hunk 5 above)
Comment (datadanne) [reply src/transformers/_common/assetTransfers.ts:258]: Do you need to call them individually somewhere? If not, the `UsabilityTransformer` is that right? A function that accepts a `RawBlock` and runs all transformers on the block and all transactions

File: src/transformers/_common/accountAbstraction/accounts/unknown.ts (hunk 2 above)
Comment (datadanne) [reply src/transformers/_common/accountAbstraction/accounts/unknown.ts:5]: I was looking at transactions to the EntryPoint contract and saw that many of them were from accounts using this these signatures. This is how I found all the account implementations actually,
1) Look at transactions to the EntryPoint
2) Reverse lookup the function signature using tenderly or openchain
3) Search for the function signature on github/google

For this one I couldn't find the source code anywhere

File: src/helpers/utils.ts [hunk 8]
```diff
@@ -196,12 +198,41 @@ export const convertDate = (epochTime: number): string => {
...
+
+export type BlockTransformer = (block: RawBlock) => RawBlock;
+
+export const isRawTransaction = (
+  v: PartialTransaction,
+): v is RawTransaction => {
+  return 'hash' in v;
+};
+
+const isTxnTransformer = (
```
Comment (pcowgill) [inline src/helpers/utils.ts:214]:
```suggestion
// Differentiating between the two types depends on the assumption that
// TxnTransformer accepts two arguments and BlockTransformer one
// If that changes, we'll need to update this
const isTxnTransformer = (
```

File: src/transformers/_common/accountAbstraction/accounts/unknown.ts (hunk 2 above)
Comment (pcowgill) [reply src/transformers/_common/accountAbstraction/accounts/unknown.ts:5]:
```suggestion
// Source unknown
// To find this, looked at transactions to the EntryPoint contract
// and saw that many of them were from accounts using this these signatures.

// Method for finding them:
// - Look at transactions to the EntryPoint
// - Reverse lookup the function signature using tenderly or openchain
// - Search for the function signature on github/google

// For this one, couldn't find the source code anywhere
```

File: src/transformers/_common/assetTransfers.ts (hunk 5 above)
Comment (pcowgill) [reply src/transformers/_common/assetTransfers.ts:258]: From a discussion outside of GitHub: We don't have the export for the individual ones, but we do still have an export at the top level with block in, block out

File: src/transformers/_common/assetTransfers.ts (hunk 7 above)
Comment (pcowgill) [reply src/transformers/_common/assetTransfers.ts:175]: Yes, this makes sense, thanks!

File: src/transformers/_common/sigHash.ts (hunk 6 above)
Comment (pcowgill) [reply src/transformers/_common/sigHash.ts:5]: If you find anything based on this investigation, we can circle back in a follow-up PR. This doesn't need to block merging.
